### DISCUSSION TITLE
Taking gradient with respect to only some of a model's weights

### DISCUSSION THREAD

**Discussion by newalexander** [google/flax#2101] [Q&A] [answered]
Hello!

I'm interested in taking the gradient of a loss function with respect to only a subset of a neural network's weights. 

I have an attempt at implementing this below -- the key aspects are splitting apart the parameters at the beginning of `grad_fn` and then re-joining them in `_loss`. Is my solution the best way to handle this use-case? Or is there a more natural way?

Thank you!

```python
from jax import random, grad
from flax.core.frozen_dict import FrozenDict
from flax import linen as nn


class MLP(nn.Module):
    """simple network example"""
    @nn.compact
    def __call__(self, x):
        x = nn.Dense(32, name="dense1")(x)
        x = nn.relu(x)
        x = nn.Dense(32, name="dense2")(x)
        return x


def grad_fn(_p):
    # split apart the network's parameters by layer
    p_1 = {'dense1': _p['params']['dense1']}
    p_2 = {'dense2': _p['params']['dense2']}

    def _loss(_p_0, _p_else):
        # join together the parameters into a single FrozenDict
        params = FrozenDict({'params': {**_p_0, **_p_else}})
        return (MLP().apply(params, x_sample) ** 2).sum()

    _grad_fn = grad(_loss, argnums=0)  # only take gradient wrt _p_0

    return _grad_fn(p_1, p_2)


if __name__ == '__main__':
    key_p, key_x = random.split(random.PRNGKey(0))
    x_sample = random.normal(key_x, (12, 4))
    p = MLP().init(key_p, x_sample)

    grad_fn(p)

```

**Comment by marcvanzee** [comment, marked as answer]
Hi @newalexander, please take a look at #1453, which contains quite an elaborate answer.
